File: algorithms/3_hard/M/00689_maximum_sum_of_3_non_overlapping_subarrays/prefix.c
struct entity {
    int sum;
    int idx;
};

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* maxSumOfThreeSubarrays(const int *restrict nums, const int len, const int k, int *restrict return_len) {
    if (k < 1 || k > len / 3 || len > 20000 || nums == NULL || return_len == NULL) {
        goto err;
    }
    int *restrict prefix_sum = malloc((size_t)(len + 1) * sizeof(*prefix_sum));
    if (prefix_sum == NULL) {
        goto err;
    }
    prefix_sum[0] = 0;
    for (int i = 0; i < len; i += 1) {
        if (nums[i] < 1 || nums[i] >= (1 << 16)) {
            exit(1);
        }
        prefix_sum[i + 1] = prefix_sum[i] + nums[i];
    }

    struct entity *restrict buff = calloc((size_t)(len + 1) * 2, sizeof(*buff));
    if (buff == NULL) {
        goto err_clean_prefix;
    }
    struct entity *restrict prefix_max = &buff[0];
    struct entity *restrict suffix_max = &buff[len + 1];

    for (int i = 0; i <= len - k; i += 1) {
        const int curr = prefix_sum[i + k] - prefix_sum[i];
        if (i == 0 || curr > prefix_max[i - 1].sum) {
            prefix_max[i] = (struct entity){ .sum = curr, .idx = i };
        } else {
            prefix_max[i] = prefix_max[i - 1];
        }
    }

    for (int i = len - k; i >= 0; i -= 1) {
        const int curr = prefix_sum[i + k] - prefix_sum[i];
        if (curr >= suffix_max[i + 1].sum) {
            suffix_max[i] = (struct entity){ .sum = curr, .idx = i };
        } else {
            suffix_max[i] = suffix_max[i + 1];
        }
    }

    int sum = 0;
    int* res = calloc(3, sizeof(*res));
    if (res == NULL) {
        goto err_clean_buff;
    }

    for (int i = k; i <= len - k - k; i += 1) {
        const int curr_sum = prefix_sum[i + k] - prefix_sum[i] +
                             prefix_max[i - k].sum + 
                             suffix_max[i + k].sum;

        if (curr_sum > sum) {
            sum = curr_sum;
            res[0] = prefix_max[i - k].idx;
            res[1] = i;
            res[2] = suffix_max[i + k].idx;
        }
    }

    *return_len = 3;
    free(buff);
    free(prefix_sum);
    return res;

err_clean_buff:
    free(buff);
err_clean_prefix:
    free(prefix_sum);
err:
    *return_len = 0;
    return NULL;
}

File: algorithms/3_hard/M/00689_maximum_sum_of_3_non_overlapping_subarrays/prefix.rs
impl Solution {
    pub fn max_sum_of_three_subarrays(nums: Vec<i32>, k: i32) -> Vec<i32> {
        assert!(k >= 1);
        let k = k as usize;
        let len = nums.len();
        assert!(k <= len / 3 && len <= 20_000);

        let mut prefix_sum = vec![0_i32; len + 1];
        for i in 0..len {
            assert!(nums[i] >= 1 && nums[i] <= 2_i32.pow(16));
            prefix_sum[i + 1] = prefix_sum[i] + nums[i];
        }

        let mut buff = vec![(0_i32, 0_i32); (len + 1) * 2];
        let (mut prefix_max, mut suffix_max) = buff.split_at_mut(len + 1);

        for i in 0..=(len - k) {
            let curr = prefix_sum[i + k] - prefix_sum[i];
            if i == 0 || curr > prefix_max[i - 1].0 {
                prefix_max[i] = (curr, i as i32);
            } else {
                prefix_max[i] = prefix_max[i - 1];
            }
        }

        for i in (0..=(len - k)).rev() {
            let curr = prefix_sum[i + k] - prefix_sum[i];
            if curr >= suffix_max[i + 1].0 {
                suffix_max[i] = (curr, i as i32);
            } else {
                suffix_max[i] = suffix_max[i + 1];
            }
        }

        let mut sum = 0_i32;
        let mut res = vec![0_i32; 3];

        for i in k..=(len - k - k) {
            let curr_sum = prefix_sum[i + k] - prefix_sum[i] +
                           prefix_max[i - k].0 + 
                           suffix_max[i + k].0;

            if curr_sum > sum {
                sum = curr_sum;
                (res[0], res[1], res[2]) = (prefix_max[i - k].1, i as i32, suffix_max[i + k].1); 
            }
        }

        res
    }
}
